### PR TITLE
Updates for a new database

### PR DESCRIPTION
**Pull request by 0xf8f8ff**
- new and updated protobuf messages
- one step closer to replacing all UI models
- faster and easier fetching of all student or group submissions for lab result pages
- new database constraint on student and group repository entries

### CODE REVIEW

File: ag/ag.proto [hunk 1]
```diff
@@ -2,6 +2,9 @@ syntax = "proto3";
 
 import "github.com/gogo/protobuf/gogoproto/gogo.proto";
 
+//  [(gogoproto.moretags) = "sql:\"-\""] tag marks fields that will be ignored by the database
+// unique_index tags set SQL UNIQUE constraint for the database
+
```
[the remark this answers is not in the document]
Comment (0xf8f8ff) [reply ag/ag.proto:7]: I really like the idea, but somehow just can't get imports to work properly, which makes it impossible to compile. Working on it.

[the remark this answers is not in the document]
Comment (meling) [reply ag/ag.proto:7]: I think you only need to make sure that the import path is defined with the `-I` flag.

File: ag/ag.proto [hunk 2]
```diff
@@ -166,14 +168,21 @@ message Submissions {
     repeated Submission submissions = 1;
 }
 
-message LabResultLink {    
-    string authorName = 1;
-    Enrollment enrollment = 2;
-    repeated Submission submissions = 3;
+//   UI structures, never saved in the database   //
+
+message SubmissionLink {
```
[the remark this answers is not in the document]
Comment (0xf8f8ff) [reply ag/ag.proto:173]: `SubmissionLink` and `EnrollmentLink` are destined to replace `IStudentLabsForCourse` and `IStudentLab` UI models, so yes, they will be used in the browser by themselves. They are just waiting for a proper marshalling of the `Submission` message fields in order to stop using `ISubmission` model, then the rest of UI models will be replaced by the protobuf messages completely.

File: web/courses.go [hunk 3]
```diff
@@ -125,38 +124,96 @@ func (s *AutograderService) getSubmissions(request *pb.SubmissionRequest) (*pb.S
...
+	for _, grp := range course.Groups {
+
+		newLink := &pb.EnrollmentLink{}
+		for _, enrol := range course.Enrollments {
+			if enrol.GroupID > 0 && enrol.GroupID == grp.ID {
+				newLink.Enrollment = enrol
+			}
+		}
+		if newLink.Enrollment == nil {
+			return nil, fmt.Errorf("Got empty enrollment for group %+v", grp)
```
[the remark this answers is not in the document]
Comment (0xf8f8ff) [reply web/courses.go:198]: I suppose it is sufficient to just log, and not return any errors at all. Empty slice of `enrolLinks` is legal (course can have no assignments or no groups without causing `Results` and `GroupResults` UI problems ) and should not return any errors.

Remover error there, added logging.

File: envoy/envoy.Dockerfile [hunk 4]
```diff
@@ -1,3 +1,3 @@
-FROM envoyproxy/envoy:latest
+FROM envoyproxy/envoy:v1.14.1
```
[the remark this answers is not in the document]
Comment (0xf8f8ff) [reply envoy/envoy.Dockerfile:1]: `latest` tag is discontinued by the Envoy team: https://github.com/envoyproxy/envoy/pull/10800/files

File: web/courses.go [hunk 5]
```diff
@@ -341,56 +393,3 @@ func (s *AutograderService) enrollTeacher(ctx context.Context, sc scm.SCM, enrol
...
-	if groupLab {
-		for _, grp := range course.Groups {
-			groupSubmissions := make([]*pb.Submission, 0)
-			for _, v := range labCache[grp.GetID()] {
-				tmp := v
-				groupSubmissions = append(groupSubmissions, &tmp)
-			}
-
-			// sort latest submissions by lab ID
-			sort.Slice(groupSubmissions, func(i, j int) bool {
```
[the remark this answers is not in the document]
Comment (0xf8f8ff) [reply web/courses.go:357]: Yeah, the sort is still useful. Added back.

File: web/courses.go [hunk 6]
```diff
@@ -341,56 +393,3 @@ func (s *AutograderService) enrollTeacher(ctx context.Context, sc scm.SCM, enrol
...
-		for _, usr := range course.Enrollments {
-			// collect all submission values for the user
-			userSubmissions := make([]*pb.Submission, 0)
-			for _, v := range labCache[usr.GetUserID()] {
-				tmp := v
-				userSubmissions = append(userSubmissions, &tmp)
-			}
-
-			// sort latest submissions by lab ID
-			sort.Slice(userSubmissions, func(i, j int) bool {
```
[the remark this answers is not in the document]
Comment (0xf8f8ff) [reply web/courses.go:382]: Fixed.

File: database/gormdb_course.go [hunk 7]
```diff
@@ -56,7 +56,7 @@ func (db *GormDB) GetCourse(courseID uint64, withInfo bool) (*pb.Course, error)
 		}
 		// and only group submissions from approved groups
 		modelGroup := &pb.Group{Status: pb.Group_APPROVED, CourseID: courseID}
-		if err := m.Preload("Assignments").Preload("Enrollments", "status in (?)", userStates).Preload("Enrollments.User").Preload("Groups", modelGroup).First(&course, courseID).Error; err != nil {
+		if err := m.Preload("Assignments").Preload("Enrollments", "status in (?)", userStates).Preload("Enrollments.User").Preload("Enrollments.Group").Preload("Groups", modelGroup).First(&course, courseID).Error; err != nil {
```
[the remark this answers is not in the document]
Comment (0xf8f8ff) [reply database/gormdb_course.go:59]: Done.

File: ag/ag.proto [hunk 8]
```diff
@@ -260,11 +269,20 @@ message Status {
     string Error = 2;
 }
 
-message LabRequest {
+message SubmissionLinkRequest {
```
[the remark this answers is not in the document]
Comment (0xf8f8ff) [reply ag/ag.proto:272]: Renamed.

File: web/courses.go [hunk 9]
```diff
@@ -125,38 +125,99 @@ func (s *AutograderService) getSubmissions(request *pb.SubmissionRequest) (*pb.S
...
+	for _, grp := range course.Groups {
+
+		newLink := &pb.EnrollmentLink{}
+		for _, enrol := range course.Enrollments {
+			if enrol.GroupID > 0 && enrol.GroupID == grp.ID {
+				newLink.Enrollment = enrol
+			}
+		}
+		if newLink.Enrollment == nil {
+			s.logger.Debugf("Got empty enrollment for group %+v", grp)
```
Comment (meling) [inline web/courses.go:201]: Okay, this is fine, but this method no longer needs to return an error, making it easier to use the result in `getAllLabs`.

**Pull request merged**